### PR TITLE
Add headless oauth support

### PR DESCRIPTION
**Pull request by sqrrrl**
Adds --no-localhost option to login command + refactors auth a little.

Planning a subsequent change to upgrade the client lib & auth library to the recently released version (v25.x for client, v1.x) since the new auth library has support for code_verifier in the oauth flow which will makes things more secure.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -40,6 +40,9 @@ const recursive = require('recursive-readdir');
 const Spinner = require('cli-spinner').Spinner;
 const splitLines = require('split-lines');
 const url = require('url');
+const readline = require('readline');
+
```
Comment (grant) [inline index.js]: Remove \n.

File: index.js [hunk 2]
```diff
@@ -249,6 +251,100 @@ function getAPICredentials(cb) {
...
+/**
+ * Requests authorization to manage Apps Scrpit projects. Spins up
+ * a temporary HTTP server to handle the auth redirect.
+ *
+ * @param {Object} opts OAuth2 options
+ * @return {Promise} Promise resolving with the authorization code
+ */
+function authorizeWithLocalhost(opts) {
+  return new Promise((resolve, reject) => {
+    var server = http.createServer(function(req, res) {
```
Comment (grant) [inline index.js]: Use `(req, res) =>`.

File: index.js [hunk 3]
```diff
@@ -249,6 +251,100 @@ function getAPICredentials(cb) {
...
+      var url_parts = url.parse(req.url, true);
```
Comment (grant) [inline index.js]: Use `urlParts`. No snake case. Here and elsewhere.

File: index.js [hunk 4]
```diff
@@ -249,6 +251,100 @@ function getAPICredentials(cb) {
...
+    authorizeWithoutLocalhost(opts);
+
+  authCode.then(code => {
+    return new Promise((res, rej) => {
+      oauth2Client.getToken(code, (err, token) => {
+        if (err) {
+          reject(err);
+        } else {
+          res(token);
+        }
```
Comment (grant) [inline index.js]: Let's shorten when possible:

```js
if (err) return reject(err);
return res(token);
```

File: index.js [hunk 5]
```diff
@@ -249,6 +251,100 @@ function getAPICredentials(cb) {
   });
 }
 
+/**
+ * Requests authorization to manage Apps Scrpit projects.
+ * @param {Boolean} useLocalhost True if a local HTTP server should be run
+ *     to handle the auth response. False if manual entry used.
+ */
+ function authorize(useLocalhost) {
+  var opts = {
```
Comment (grant) [inline index.js]: Use `let` instead of `var` everywhere.

File: index.js [hunk 6]
```diff
@@ -249,6 +251,100 @@ function getAPICredentials(cb) {
...
+ * used.
+ *
+ * @param {Object} opts OAuth2 options
+ * @return {Promise} Promise resolving with the authorization code
+ */
+
+function authorizeWithoutLocalhost(opts) {
+  oauth2Client._redirectUri = REDIRECT_URI_OOB;
+  var authUrl = oauth2Client.generateAuthUrl(opts);
+
```
Comment (grant) [inline index.js]: Remove \n

File: index.js [hunk 7]
```diff
@@ -249,6 +251,100 @@ function getAPICredentials(cb) {
...
+  var opts = {
+    access_type: 'offline',
+    scope: [
+      'https://www.googleapis.com/auth/script.deployments',
+      'https://www.googleapis.com/auth/script.projects',
+    ],
+  }
+  var authCode = useLocalhost ?
+    authorizeWithLocalhost(opts) :
+    authorizeWithoutLocalhost(opts);
```
Comment (grant) [inline index.js]: Great ternary!

File: index.js [hunk 8]
```diff
@@ -249,6 +251,100 @@ function getAPICredentials(cb) {
...
+      if (url_parts.query.code) {
+        resolve(url_parts.query.code);
+      } else {
+        reject(url_parts.query.error);
+      }
+      res.end(LOG.AUTH_PAGE_SUCCESSFUL);
+      server.shutdown();
+    }).withShutdown();
+
+    server.listen(0, () => {
```
Comment (grant) [inline index.js]: Listen on port 0? I assume it's not in use. o.w. we'll get `ERR_SERVER_ALREADY_LISTEN`.

**Comment by grant** [comment]
Fixes #2.

File: index.js (hunk 8 above)
Comment (sqrrrl) [reply index.js]: 0 just means pick an unused ephemeral port. For installed app flows, port is ignored for registered callback localhost URLs, specifically for this reason -- hardcoded ports can cause conflicts if there's already a service running on it.

File: index.js (hunk 2 above)
Comment (sqrrrl) [reply index.js]: done

File: index.js (hunk 6 above)
Comment (sqrrrl) [reply index.js]: done

File: index.js (hunk 4 above)
Comment (sqrrrl) [reply index.js]: done

File: index.js (hunk 5 above)
Comment (sqrrrl) [reply index.js]: done

File: index.js (hunk 1 above)
Comment (sqrrrl) [reply index.js]: done

**Comment by sqrrrl** [comment]
PTAL

**Pull request merged**